### PR TITLE
Secure hash-based password authentication

### PR DESCRIPTION
**Pull request by JorgeFeijoom**
Register, password recovery, login and profile edition hash-based

### CODE REVIEW

**Comment by lortschi** [comment]
looks all fine to me

File: config/server.production.js [hunk 1]
```diff
@@ -32,16 +32,7 @@ module.exports = {
...
-	smtpServer: {
-		host: '',
-		port: 0,
-		secure: true,
-		user: '',
-		pass: '',
-		fromName: '',
-		fromAddress: ''
-	},
+	smtpServer: {},
```
Comment (ngoyal16) [inline config/server.production.js]: smtpServer setting should be back

File: src/ajaxRouter.js [hunk 2]
```diff
@@ -139,57 +140,56 @@ ajaxRouter.get('/cart', (req, res, next) => {
...
+		};
 
-	// update customer password after checking customer id
-	if ('id' in req.body) {
-		await api.customers
-			.update(userId, customerDraft)
-			.then(({ status, json }) => {
+		// update customer password after checking customer id
+		if ('id' in req.body) {
+			api.customers.update(userId, customerDraft).then(({ status, json }) => {
```
Comment (ngoyal16) [inline src/ajaxRouter.js]: use async/await for better managment

File: src/ajaxRouter.js [hunk 3]
```diff
@@ -139,57 +140,56 @@ ajaxRouter.get('/cart', (req, res, next) => {
...
+			return false;
+		}
 
-	// if customer email exists send status back
-	await api.customers.list(filter).then(({ status, json }) => {
-		if (json.total_count > 0) {
-			data.status = true;
-			data.id = AuthHeader.encodeUserLoginAuth(userId);
+		// if customer email exists send status back
+		api.customers.list(filter).then(({ status, json }) => {
```
Comment (ngoyal16) [inline src/ajaxRouter.js]: put asyn/await before promise function else it will not wait for results

File: src/ajaxRouter.js [hunk 4]
```diff
@@ -305,25 +308,35 @@ ajaxRouter.post('/login', async (req, res, next) => {
...
-
-					let objJsonB64 = JSON.stringify(customerData);
-					objJsonB64 = Buffer.from(objJsonB64).toString('base64');
-					res.status(status).send(JSON.stringify(objJsonB64));
-				});
+			bcrypt.compare(inputPassword, customerPassword, function(err, out) {
+				if (out == true) {
+					customerData.token = AuthHeader.encodeUserLoginAuth(result._id);
+					customerData.authenticated = true;
+					api.customers.retrieve(result._id).then(({ status, json }) => {
```
Comment (ngoyal16) [inline src/ajaxRouter.js]: async function when doing any call which is should be require in same thread

File: src/index.js [hunk 5]
```diff
@@ -33,14 +33,24 @@ app.use(express.static('public/content', STATIC_OPTIONS));
...
+	//var allowedOrigins = security.getAccessControlAllowOrigin();
+	//var origin = req.headers.origin;
+	//if (allowedOrigins === '*') {
+	//res.setHeader('Access-Control-Allow-Origin', allowedOrigins);
+	//} else {
+	//if (allowedOrigins.indexOf(origin) > -1) {
+	//res.setHeader('Access-Control-Allow-Origin', origin);
+	//}
+	//}
+	var allowedOrigins = [
```
Comment (ngoyal16) [inline src/index.js]: do not hardcore cors header.. it should come to from config file.

**Review by ngoyal16 — CHANGES_REQUESTED** [review]

File: src/ajaxRouter.js [hunk 6]
```diff
@@ -10,9 +10,10 @@ import AuthHeader from './lib/auth-header';
 import mailer from './lib/mailer';
 import EmailTemplatesService from './services/settings/emailTemplates';
 import SettingsService from './services/settings/settings';
+import bcrypt from 'bcrypt';
 
+const saltRounds = 10;
```
Comment (LuBuss) [inline src/ajaxRouter.js]: Salt should be config option

**Review by LuBuss — COMMENTED** [review]
Agree with all @ngoyal16 comments

**Comment by JorgeFeijoom** [comment]
- fixed hardcore CORS header
- smtp parameters on server.production.js
- set salt rounds on server conf
- fixed async/await on api calls and better management

**Review by vamcart — APPROVED** [review]

**Review by ngoyal16 — APPROVED** [review]
seems good to go

**Review by ngoyal16 — APPROVED** [review]

**Pull request merged**